Source/Core/Core/PowerPC/Expression.cpp

```diff
@@ -0,0 +1,229 @@
+                                   {"u32", CastFunc<u32>},
+                                   {"s32", CastFunc<s32, u32>},
+                                   {}};
+
+const std::unordered_map<std::string_view, u32&> g_register_vars = {
+    {"r0", GPR(0)},   {"r1", GPR(1)},   {"r2", GPR(2)},   {"r3", GPR(3)},   {"r4", GPR(4)},
+    {"r5", GPR(5)},   {"r6", GPR(6)},   {"r7", GPR(7)},   {"r8", GPR(8)},   {"r9", GPR(9)},
+    {"r10", GPR(10)}, {"r11", GPR(11)}, {"r12", GPR(12)}, {"r13", GPR(13)}, {"r14", GPR(14)},
+    {"r15", GPR(15)}, {"r16", GPR(16)}, {"r17", GPR(17)}, {"r18", GPR(18)}, {"r19", GPR(19)},
+    {"r20", GPR(20)}, {"r21", GPR(21)}, {"r22", GPR(22)}, {"r23", GPR(23)}, {"r24", GPR(24)},
+    {"r25", GPR(25)}, {"r26", GPR(26)}, {"r27", GPR(27)}, {"r28", GPR(28)}, {"r29", GPR(29)},
+    {"r30", GPR(30)}, {"r31", GPR(31)}, {"pc", PC},       {"lr", LR},       {"ctr", CTR},
+};
+
+const std::unordered_map<std::string_view, u64&> g_float_register_vars = {
+    {"f0", rPS(0).ps0},   {"f1", rPS(1).ps0},   {"f2", rPS(2).ps0},   {"f3", rPS(3).ps0},
+    {"f4", rPS(4).ps0},   {"f5", rPS(5).ps0},   {"f6", rPS(6).ps0},   {"f7", rPS(7).ps0},
+    {"f8", rPS(8).ps0},   {"f9", rPS(9).ps0},   {"f10", rPS(10).ps0}, {"f11", rPS(11).ps0},
+    {"f12", rPS(12).ps0}, {"f13", rPS(13).ps0}, {"f14", rPS(14).ps0}, {"f15", rPS(15).ps0},
+    {"f16", rPS(16).ps0}, {"f17", rPS(17).ps0}, {"f18", rPS(18).ps0}, {"f19", rPS(19).ps0},
+    {"f20", rPS(20).ps0}, {"f21", rPS(21).ps0}, {"f22", rPS(22).ps0}, {"f23", rPS(23).ps0},
+    {"f24", rPS(24).ps0}, {"f25", rPS(25).ps0}, {"f26", rPS(26).ps0}, {"f27", rPS(27).ps0},
+    {"f28", rPS(28).ps0}, {"f29", rPS(29).ps0}, {"f30", rPS(30).ps0}, {"f31", rPS(31).ps0},
+};
+
+enum class SynchronizeDirection
+{
+  FromRegister,
+  ToRegister,
+};
+
+static void SynchronizeRegisters(const expr_var_list& vars, SynchronizeDirection dir)
+{
+  for (auto* v = vars.head; v != nullptr; v = v->next)
+  {
+    auto r = g_register_vars.find(v->name);
+    if (r != g_register_vars.end())
+    {
+      if (dir == SynchronizeDirection::FromRegister)
+      {
+        v->value = static_cast<double>(r->second);
+      }
+      else
+      {
+        r->second = static_cast<u32>(static_cast<s64>(v->value));
+      }
+      continue;
+    }
+
+    auto fr = g_float_register_vars.find(v->name);
+    if (fr != g_float_register_vars.end())
+    {
+      if (dir == SynchronizeDirection::FromRegister)
+      {
+        v->value = Common::BitCast<double>(fr->second);
+      }
+      else
+      {
+        fr->second = Common::BitCast<u64>(v->value);
+      }
+      continue;
+    }
+
+    if (dir == SynchronizeDirection::FromRegister)
+    {
+      v->value = 0;  // init expression local variables to zero
+    }
+  }
+}
+
+void ExprDeleter::operator()(expr* expression) const
+{
+  expr_destroy(expression, nullptr);
+}
+
+void ExprVarListDeleter::operator()(expr_var_list* vars) const
+{
+  expr_destroy(nullptr, vars);  // free list elements
```

Comment:
We place documentation comments above the line they apply to rather than off to the side where applicable.